scipy/cluster/hierarchy.py

```diff
@@ -1391,6 +1391,170 @@ def cut_tree(Z, n_clusters=None, height=None):
+    consideration. If the cluster contains a number of samples lower than the
+    given maximum, it is taken as result and its offspring sub-clusters not
+    further processed.
+
+    Parameters
+    ----------
+    Z : scipy.cluster.linkage array
+        The linkage matrix resulting (for instance) from calling the method
+        scipy.cluster.hierarchy.ward(). I.e. it contains the hierarchical
+        clustering encoded as a linkage matrix.
+
+    max_cluster_size : array_like
+        maximum number of data samples contained within the resulting clusters.
+        Thus, all resulting clusters will contain a number of data samples
+        <= max_cluster_size. Note that max_cluster_size must be >= 1.
+
+    Returns
+    -------
+    cluster_id : ndarray
+        one-dimensional numpy array of integers containing for each input
+        sample its corresponding cluster id. The cluster id is an integer which
+        is higher for deeper tree levels.
+
+    cluster_level : ndarray
+        one-dimensional numpy array of strings containing for each input sample
+        its corresponding cluster tree level, i.e. a string of '0's and '1's
+        separated by '.' Note that the cluster level is longer for deeper tree
+        levels, being 0 the root cluster, 0.0 and 0.1 its offspring, and so on.
+
+    Examples
+    --------
+    >>> from scipy import cluster
+    >>> from scipy import stats
+
+    Initialize the random seed.
+
+    >>> np.random.seed(14)
+
+    Create a input matrix containing 100 data samples with 4 dimensions. Note:
+    using gamma() in order to generate an unbalanced distribution. If a regular
+    cut_tree() would be performed, one big and many small clusters would be
+    obtained.
+
+    >>> X = stats.gamma.rvs(0.1, size=400).reshape((100,4))
+
+    Compute the linkage matrix using the scipy ward() method.
+
+    >>> Z = cluster.hierarchy.ward(X)
+
+    Perform a balanced cut tree of the linkage matrix.
+
+    >>> [cluster_id, cluster_level] = cut_tree_balanced(Z, max_cluster_size=10)
+    >>> cluster_id[:10]
+    >>> cluster_level[:10]
+    """
+    try:
+        # Assert that the input max_cluster_size is >= 1
+        assert max_cluster_size >= 1
+
+        # Perform a full cut tree of the linkage matrix
+        full_cut = cut_tree(Z)
+
+        # Initialize the vble containing the current cluster id (it will be
+        # higher for each newly found valid cluster)
+        last_cluster_id = 1
+
+        # Initialize the resulting cluster id vector (containing for each row
+        # in input_data_x_sample its corresponding cluster id)
+        cluster_id = np.zeros(full_cut.shape[1], dtype=int)
+
+        # Initialize the resulting cluster level vector (containing for each
+        # data sample its corresponding cluster tree level)
+        cluster_level = np.array(['0' for _ in
+                                  range(full_cut.shape[1])], dtype=object)
```

Comment:
The mentioned structure was changed from an array of strings, to an array of integer arrays.

Note that the initialization of the structure was a little bit more challenging than expected, otherwise I got problems with numpy broadcast operations.

`cluster_level = np.empty((full_cut.shape[1],), dtype=object)`
 `for i in range(full_cut.shape[1]): cluster_level[i] = np.array([0],int)`

For more information see [this stackoverflow contribution](https://stackoverflow.com/questions/33983053/how-to-create-a-numpy-array-of-lists).